File: ObjC/ASTM30View-ObjC/ASTM30View-ObjC/ASTM30GraphicViewController/ASTM30GraphicViewController.m
//
//  ASTM30GraphicViewController.m
//  ASTM30View-ObjC
//
//  Created by Inaba Mizuki on 2016/1/26.
//  Copyright © 2016年 Inaba Mizuki. All rights reserved.
//

#import "ASTM30GraphicViewController.h"
#import "ASTM30CoordinateSpace.h"
#import "ASTM30PointsInfo.h"
#import "MZ.h"

@interface PointsInfoToLayersDictionary : NSObject
@end

@implementation PointsInfoToLayersDictionary {
    NSMutableDictionary<NSString*, ASTM30PointsInfo*>* _pointsInfosDict;
    NSMutableDictionary<NSString*, CAShapeLayer*>* _pointsLayersDict;
}

- (instancetype)init {
    self = [super init];

    _pointsInfosDict = [NSMutableDictionary dictionary];
    _pointsLayersDict = [NSMutableDictionary dictionary];

    return self;
}

- (id)objectForKeyedSubscript:(ASTM30PointsInfo *)info {
    return nil;
}

- (void)setObject:(CAShapeLayer *)layer forKeyedSubscript:(ASTM30PointsInfo *)info {

}

- (void)dealloc {
    [_pointsInfosDict removeAllObjects];
    [_pointsLayersDict removeAllObjects];
}

@end



@interface ASTM30GraphicViewController (PresentViewsAndLayers)
- (void)_setAndAddGraphicBackgroundViewToView:(UIView *)view;
- (void)_setAndAddGridLayerToView:(UIView *)view;
- (void)_setAndAddPointsLayersViewToView:(UIView *)view;
- (void)_setAndAddAllPoinsInfoLayersToView:(UIView *)view;
- (void)_setGraphicBackgroundMaskWithPointsInfoName:(NSString * _Nullable)name;
- (void)_setAndAddReferenceToTestSourceArrowsLayerToView:(UIView *)view;
- (CAShapeLayer *)_shapeLayerWithPointsInfo:(ASTM30PointsInfo *)pointsInfo;
- (CAShapeLayer *)_maskLayerFromLayer:(CAShapeLayer *)layer;
- (UIBezierPath *)_arrowPathFromPoint:(CGPoint)from toPoint:(CGPoint)to;
@end

@interface ASTM30GraphicViewController (ASTM30GraphicTypeSwitchAnimations)
- (void)_animateMaskEnable:(BOOL)maskEnable duration:(NSTimeInterval)duration;
- (void)_animateMaskEnableToGraphicBackground:(BOOL)maskEnable duration:(NSTimeInterval)duration;
- (void)_animateMaskEnableToGraphicBackgroundForFade:(BOOL)maskEnable duration:(NSTimeInterval)duration;
@end

@interface ASTM30GraphicViewController (Supports)
- (CGPoint)_pointFrom:(CGPoint)point inCoordinateSpace:(ASTM30CoordinateSpace*)coordinateSpace;
- (void)_addPointsInfoToDict:(ASTM30PointsInfo *)info;
- (void)_addShaperLayerToDict:(CAShapeLayer *)layer withInfo:(ASTM30PointsInfo *)info;
- (ASTM30PointsInfo * _Nullable)_getPointsInfoWithName:(NSString *)name;
- (CAShapeLayer * _Nullable)_getShapeLayerInDictWithPointsInfo:(ASTM30PointsInfo *)info;
@end



@implementation ASTM30GraphicViewController {
    NSMutableDictionary<NSString*, ASTM30PointsInfo*>* _pointsInfosDict;
    NSMutableDictionary<NSString*, CAShapeLayer*>* _pointsLayersDict;

    UIView* _pointsLinesLayerView;

    CAShapeLayer* _sourceToReferenceArrowsLayer;

    UIImageView* _graphicBackgroundView;

    CAShapeLayer* _graphicBackgroundGridLayer;

    UIImageView* _graphicBackgroundForFadeView;

    CAShapeLayer* _graphicBackgroundMaskLayer;
}

@synthesize graphicType;

- (void)viewDidLoad {
    [super viewDidLoad];
    [self _initSetting];
    self.view.backgroundColor = [UIColor blackColor];
}

- (void)viewWillLayoutSubviews {
    [super viewWillLayoutSubviews];
    [self refresh];
}

- (void)addPointsInfo:(ASTM30PointsInfo * _Nonnull)info {
    [[self _getShapeLayerInDictWithPointsInfo:info] removeFromSuperlayer];
    if (_pointsLayersDict[info.name] != nil ) [_pointsLayersDict removeObjectForKey:info.name];

    [self _addPointsInfoToDict:info];
}

- (ASTM30PointsInfo * _Nullable)poinsInfoWithName:(NSString * _Nonnull)name {
    for (NSString* infoName in _pointsLayersDict.allKeys) {
        if ([infoName isEqualToString:name]) {
            return [self _getPointsInfoWithName:infoName];
        }
    }

    return nil;
}

- (void)removePointsInfoWithName:(NSString * _Nonnull)name {
    [_pointsInfosDict removeObjectForKey:name];
    [_pointsLayersDict removeObjectForKey:name];
}

- (void)removeAllPointsInfo {
    for (CAShapeLayer* layer in _pointsLayersDict.allValues) {
        [layer removeFromSuperlayer];
    }

    [_pointsLayersDict removeAllObjects];
}

- (void)setGraphicType:(ASTM30GraphicType)type animated:(bool)animated duration:(CFTimeInterval)duration {
    graphicType = type;
    bool enableMask = graphicType == ASTM30GraphicType_ColorDistortion;

    if (animated) {
        [self _animateMaskEnable:enableMask duration: duration];
    } else {
        [self _setGraphicBackgroundWithMaskEnable:enableMask];
    }
}

- (void)setGraphicType:(ASTM30GraphicType)type {
    [self setGraphicType:type animated:true duration:0.25];
}

- (void)refresh {
    [self _setAndAddGraphicBackgroundViewToView:self.view];
    [self _setAndAddGridLayerToView:_graphicBackgroundView];
    [self _setAndAddPointsLayersViewToView:self.view];
    [self _setAndAddAllPoinsInfoLayersToView:_pointsLinesLayerView];
    [self _setGraphicBackgroundMaskWithPointsInfoName:self.testSourceName];
//    [self _setAndAddReferenceToTestSourceArrowsLayerToView:_pointsLinesLayerView];
}


# pragma mark - Private

- (void)_initSetting {
    _pointsInfosDict = [NSMutableDictionary dictionary];
    _pointsLayersDict = [NSMutableDictionary dictionary];
    _sourceToReferenceArrowsLayer = nil;
    _graphicBackgroundView = nil;
    _graphicBackgroundGridLayer = nil;
    _graphicBackgroundForFadeView = nil;
    _graphicBackgroundMaskLayer = nil;
}

- (void)_setGraphicBackgroundWithMaskEnable:(bool)maskEnable {
    mz_guard_let_return(backgroundView, _graphicBackgroundView);
    mz_guard_let_return(backgroundMaskLayer, _graphicBackgroundMaskLayer);

    backgroundView.layer.mask = (maskEnable)? backgroundMaskLayer : nil;

    [_pointsInfosDict.allKeys forEachWithAction:^(NSString* name) {
        mz_gen_var(info, [self _getPointsInfoWithName:name]);
        mz_gen_var(shapeLayer, [self _getShapeLayerInDictWithPointsInfo:info]);

        UIColor* nextColorValue = (maskEnable)? info.colorInMasked : info.color;
        shapeLayer.strokeColor = nextColorValue.CGColor;
    }];

    _graphicBackgroundForFadeView.hidden = maskEnable;
}

@end

@implementation ASTM30GraphicViewController (PresentViewsAndLayers)

- (void)_setAndAddGraphicBackgroundViewToView:(UIView *)view {
    [_graphicBackgroundView removeFromSuperview];
    [_graphicBackgroundForFadeView removeFromSuperview];

    UIImageView *(^newGraphicView)(void) = ^{
        UIImageView* imageView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"ColorVectorBackground"]];
        imageView.frame = view.frame;
        imageView.contentMode = UIViewContentModeScaleToFill;
        
        return imageView;
    };

    _graphicBackgroundView = newGraphicView();
    _graphicBackgroundForFadeView = newGraphicView();

    [self.view addSubview:_graphicBackgroundForFadeView];
    [self.view addSubview:_graphicBackgroundView];
}

- (void)_setAndAddGridLayerToView:(UIView *)view {
    [_graphicBackgroundGridLayer removeFromSuperlayer];

    CGFloat numberOfGrid = 5;

    CGPoint interval = CGPointMake(view.frame.size.width/numberOfGrid,
                                   view.frame.size.height/numberOfGrid);

    UIBezierPath* path = [UIBezierPath bezierPath];

    for (int i = 1; i <numberOfGrid; i++) {
        UIBezierPath* linePath = [UIBezierPath bezierPath];

        [linePath moveToPoint:CGPointMake(0, i*interval.y)];
        [linePath addLineToPoint:CGPointMake(view.frame.size.width, i*interval.y)];

        [linePath moveToPoint:CGPointMake(i*interval.x, 0)];
        [linePath addLineToPoint:CGPointMake(i*interval.x, view.frame.size.height)];

        [path appendPath:linePath];
    }

    CAShapeLayer* gridLayer = [CAShapeLayer layer];
    gridLayer.frame = view.frame;
    gridLayer.path = path.CGPath;

    gridLayer.lineWidth = 2;
    gridLayer.strokeColor = [UIColor grayColor].CGColor;

    [view.layer addSublayer:gridLayer];

    _graphicBackgroundGridLayer = gridLayer;
}

- (void)_setAndAddPointsLayersViewToView:(UIView *)view {
    [_pointsLinesLayerView removeFromSuperview];

    mz_gen_var(pointsView, [[UIView alloc] initWithFrame:view.frame]);

    [view addSubview:pointsView];

    _pointsLinesLayerView = pointsView;
}

- (void)_setAndAddAllPoinsInfoLayersToView:(UIView *)view {
    for (CAShapeLayer* layer in _pointsLayersDict.allValues) {
        [layer removeFromSuperlayer];
    }
    [_pointsLayersDict removeAllObjects];

    for (ASTM30PointsInfo* info in _pointsInfosDict.allValues) {
        mz_gen_var(shapeLayer, [self _shapeLayerWithPointsInfo:info]);
        [self _addShaperLayerToDict:shapeLayer withInfo:info];
        [view.layer addSublayer:shapeLayer];
    }
}

- (void)_setGraphicBackgroundMaskWithPointsInfoName:(NSString * _Nullable)name {
    if (name == nil) return;
    mz_guard_let_return(graphicBackgroundView, _graphicBackgroundView);

     NSArray* pointsInfoKeyValue = [_pointsLayersDict filterWithFunc: ^BOOL(NSString* infoName, CAShapeLayer* _) {
         return [infoName isEqualToString:name];
     }].firstObject;

    MZAssert(pointsInfoKeyValue != nil, @"pointsInfo not found with name: \(name)");

    CAShapeLayer* layer = ((MZPair *)pointsInfoKeyValue).second;
    MZAssertIfNilWithMessage(layer, @"layer not found");

    graphicBackgroundView.layer.mask = nil;

    _graphicBackgroundMaskLayer = [self _maskLayerFromLayer:layer];

    self.testSourceName = name;
}

- (void)_setAndAddReferenceToTestSourceArrowsLayerToView:(UIView *)view {
    [_sourceToReferenceArrowsLayer removeFromSuperlayer];

    mz_guard_let_return(testSourceName, self.testSourceName);
    mz_guard_let_return(referenceName, self.referenceName);

    mz_guard_let_return(toInfo, [self poinsInfoWithName:self.testSourceName]);
    mz_guard_let_return(fromInfo, [self poinsInfoWithName:self.referenceName]);

    mz_gen_var(path, [UIBezierPath bezierPath]);

    for (ASTM30Point* fromPoint in fromInfo.points) {
        mz_gen_var(toPoint, [toInfo pointWithKey:fromPoint.key]);

        MZAssertIfNilWithMessage(toPoint, @"Can not found point with key: %@", fromPoint.key);

        mz_gen_var(modifiedFromPoint, [self _pointFrom:fromPoint.value inCoordinateSpace:self.coordinateSpace]);
        mz_gen_var(modifiedToPoint, [self _pointFrom:toPoint.value inCoordinateSpace:self.coordinateSpace]);

        [path appendPath:[self _arrowPathFromPoint:modifiedFromPoint toPoint:modifiedToPoint]];
    }

    mz_gen_var(layer, [CAShapeLayer layer]);
    layer.frame = view.frame;
    layer.path = path.CGPath;
    layer.strokeColor = [UIColor greenColor].CGColor;
    layer.lineWidth = 1;
    layer.lineCap = @"round";
    layer.lineJoin = @"bevel";
    [view.layer addSublayer:layer];

    _sourceToReferenceArrowsLayer = layer;
}


- (CAShapeLayer *)_shapeLayerWithPointsInfo:(ASTM30PointsInfo *)pointsInfo {
    NSArray* points = [pointsInfo.points mapWithFunc:^(ASTM30Point* point) {
        CGPoint realPoint = [self _pointFrom:point.value
                           inCoordinateSpace:self.coordinateSpace];

        return [NSValue valueWithCGPoint:realPoint];
    }];

    mz_gen_var(path, [UIBezierPath bezierPath]);
    [path moveToPoint:[points[0] CGPointValue]];
    for (int i = 1; i < points.count; i++) {
        [path addLineToPoint:[points[i] CGPointValue]];
    }

    if (pointsInfo.closePath) { [path closePath]; }

    mz_gen_var(layer, [CAShapeLayer layer]);
    layer.path = path.CGPath;
    layer.lineWidth = pointsInfo.lineWidth;
    layer.strokeColor = pointsInfo.color.CGColor;
    layer.fillColor = [UIColor clearColor].CGColor;

    NSDictionary* actions = [NSDictionary dictionaryWithObject:[NSNull null] forKey:@"strokeColor"];
    layer.actions = actions;

    return layer;
}

- (CAShapeLayer *)_maskLayerFromLayer:(CAShapeLayer *)layer {
    mz_gen_var(maskLayer, [[CAShapeLayer alloc] initWithLayer:layer]);
    maskLayer.frame = self.view.frame;
    maskLayer.path = layer.path;

    return maskLayer;
}


- (UIBezierPath *)_arrowPathFromPoint:(CGPoint)from toPoint:(CGPoint)to {
    return nil;
//    let lengthOfFromTo = MZ.Maths.distance(p1: from, p2: to)
//    let degreesOfFromTo = MZ.Degrees.degressFromP1(from, toP2: to)
//    let maxWingsLength = 10.cgFloatValue
//    let wingsIntervalDegrees = 30.mzFloatValue
//
//    let path = UIBezierPath()
//    let zeroLeft = CGPoint(x: -lengthOfFromTo/2, y: 0)
//    let zeroRight = CGPoint(x: lengthOfFromTo/2, y: 0)
//
//    path.moveToPoint(zeroLeft)
//    path.addLineToPoint(zeroRight)
//
//    let wingsLength = min(maxWingsLength, lengthOfFromTo/3)
//    let wingPoint1 = MZ.Maths.unitVectorFromDegrees(180-wingsIntervalDegrees)*wingsLength
//    let wingPoint2 = MZ.Maths.unitVectorFromDegrees(180+wingsIntervalDegrees)*wingsLength
//
//    path.moveToPoint(zeroRight)
//    path.addLineToPoint(zeroRight + wingPoint1)
//    path.moveToPoint(zeroRight)
//    path.addLineToPoint(zeroRight + wingPoint2)
//
//    let centerOfFromTo = CGPoint(x: (to.x + from.x)/2, y: (to.y + from.y)/2)
//    let rotationAngle = MZ.Degrees.radiansFromDegrees(degreesOfFromTo).cgFloatValue
//    path.applyTransform(CGAffineTransformMakeRotation(rotationAngle))
//    path.applyTransform(CGAffineTransformMakeTranslation(centerOfFromTo.x, centerOfFromTo.y))
//
//    return path
}

@end

@implementation ASTM30GraphicViewController (ASTM30GraphicTypeSwitchAnimations)

- (void)_animateMaskEnable:(BOOL)maskEnable duration:(NSTimeInterval)duration {
    [self _animateMaskEnableToGraphicBackground:maskEnable duration:duration];
    [self _animateMaskEnableToGraphicBackgroundForFade:maskEnable duration: duration];
//    _animateFadeMaskEnableToLayer(_graphicBackgroundGridLayer, maskEnable: maskEnable, duration: duration)
//    _animateFadeMaskEnableToLayer(_sourceToReferenceArrowsLayer, maskEnable: maskEnable, duration: duration)
//    _animateMaskEnableToPointsLinesLayer(maskEnable, duration: duration)
}

- (void)_animateMaskEnableToGraphicBackground:(BOOL)maskEnable duration:(NSTimeInterval)duration {
    mz_guard_let_return(view, _graphicBackgroundView);
    mz_guard_let_return(maskLayer, _graphicBackgroundMaskLayer);

    if (view.layer.mask == nil) {
        view.layer.mask = maskLayer;
    }

    CGFloat maxScale = 5.0;
    CGFloat minScale = 1.0;

    mz_gen_var(scale, [CAKeyframeAnimation animationWithKeyPath:@"transform.scale"]);
    scale.duration = duration;
    scale.keyTimes = @[@0.0, @1.0];
    scale.values = @[[NSNumber numberWithFloat:((maskEnable)? maxScale : minScale)],
                     [NSNumber numberWithFloat:((maskEnable)? minScale : maxScale)]];
    scale.timingFunctions = @[[CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseInEaseOut]];
    scale.fillMode = kCAFillModeForwards;
    scale.removedOnCompletion = false;

    [maskLayer removeAllAnimations];
    [maskLayer addAnimation:scale forKey: @"scale"];
}

- (void)_animateMaskEnableToGraphicBackgroundForFade:(BOOL)maskEnable duration:(NSTimeInterval)duration {
    mz_guard_let_return(view, _graphicBackgroundForFadeView);

    view.alpha = (maskEnable)? 1.0 : 0.0;

    [UIView animateWithDuration:duration
                     animations:^{ view.alpha = (maskEnable)? 0.0 : 1.0; }
                     completion:^(BOOL _) { [self _setGraphicBackgroundWithMaskEnable:maskEnable]; }];
}

@end

@implementation ASTM30GraphicViewController (Supports)

- (CGPoint)_pointFrom:(CGPoint)point inCoordinateSpace:(ASTM30CoordinateSpace*)coordinateSpace {
    mz_gen_var(size, self.view.frame.size);

    mz_gen_var(realX, size.width*((point.x - coordinateSpace.xMin)/coordinateSpace.xLength));
    mz_gen_var(realY, size.height - size.height*((point.y - coordinateSpace.yMin)/coordinateSpace.yLength));

    return CGPointMake(realX, realY);
}

- (void)_addPointsInfoToDict:(ASTM30PointsInfo *)info {
    _pointsInfosDict[info.name] = info;
}

- (void)_addShaperLayerToDict:(CAShapeLayer *)layer withInfo:(ASTM30PointsInfo *)info {
    _pointsInfosDict[info.name] = info;
    _pointsLayersDict[info.name] = layer;
}

- (ASTM30PointsInfo * _Nullable)_getPointsInfoWithName:(NSString *)name {
    return _pointsInfosDict[name];
}

- (CAShapeLayer * _Nullable)_getShapeLayerInDictWithPointsInfo:(ASTM30PointsInfo *)info {
    return _pointsLayersDict[info.name];
}

@end

File: Swift/ASColorVectorViewProj/ASTM30GraphicViewController/ASTM30GraphicViewController.swift
//
//  ASTM30GraphicViewController.swift
//
//  Created by Inaba Mizuki on 2016/1/21.
//  Copyright © 2016年 AsenseTek. All rights reserved.
//

import UIKit


enum ASTM30GraphicType: Int {
    case ColorVector
    case ColorDistortion
}

class ASTM30GraphicViewController: UIViewController {

    var coordinateSpace = ASTM30CoordinateSpace()

    var testSourceName: String? = nil

    var referenceName: String? = nil

    private (set) var graphicType: ASTM30GraphicType = .ColorVector

    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = UIColor.blackColor()
    }

    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        refresh()
    }

    func addPointsInfo(info: ASTM30PointsInfo) {
        _pointsInfoToLayersDict[info]?.removeFromSuperlayer()
        _pointsInfoToLayersDict.removeValueForKey(info)

        _pointsInfoToLayersDict[info] = CAShapeLayer()
    }

    func poinsInfoWithName(name: String) -> ASTM30PointsInfo? {
        return _pointsInfoToLayersDict.keys.filter {
            info in info.name == name
        }.first
    }

    func removePointsInfoWithName(name: String) {
        guard let targetKey = poinsInfoWithName(name) else { return }

        _pointsInfoToLayersDict.removeValueForKey(targetKey)
    }

    func removeAllPointsInfo() {
        _pointsInfoToLayersDict.values.forEach { layer in layer.removeFromSuperlayer() }
        _pointsInfoToLayersDict.removeAll()
    }

    func refresh() {
        _setAndAddGraphicBackgroundViewToView(view)
        _setAndAddGridLayerToView(_graphicBackgroundView!)
        _setAndAddPointsLayersViewToView(view)
        _setAndAddAllPoinsInfoLayersToView(_pointsLinesLayerView!)
        _setGraphicBackgroundMaskWithPointsInfoName(testSourceName)
        _setAndAddReferenceToTestSourceArrowsLayerToView(_pointsLinesLayerView!)
    }

    func setGraphicType(type: ASTM30GraphicType, animated: Bool = false, duration: CFTimeInterval = 0.25) {
        graphicType = type
        let enableMask = graphicType == .ColorDistortion

        if animated {
            _animateMaskEnable(enableMask, duration: duration)
        } else {
            _setGraphicBackgroundWithMaskEnable(enableMask)
        }
    }

    
    // MARK: Private

    private var _pointsInfoToLayersDict = [ASTM30PointsInfo: CAShapeLayer]()

    private var _pointsLinesLayerView: UIView? = nil

    private var _sourceToReferenceArrowsLayer: CAShapeLayer? = nil

    private var _graphicBackgroundView: UIImageView? = nil

    private var _graphicBackgroundGridLayer: CAShapeLayer? = nil

    private var _graphicBackgroundForFadeView: UIImageView? = nil

    private var _graphicBackgroundMaskLayer: CAShapeLayer? = nil

    private func _setGraphicBackgroundWithMaskEnable(maskEnable: Bool) {
        guard let backgroundView = _graphicBackgroundView else { return }
        guard let backgroundMaskLayer = _graphicBackgroundMaskLayer else { return }

        backgroundView.layer.mask = maskEnable ? backgroundMaskLayer : nil

        _pointsInfoToLayersDict.forEach {
            (info, shapeLayer) in
            let nextColorValue = maskEnable ? info.colorInMasked : info.color

            guard let nextColor = nextColorValue else { return }

            shapeLayer.strokeColor = nextColor.CGColor
        }

        _graphicBackgroundForFadeView?.hidden = maskEnable
    }
}


//  MARK: Present views and layers
extension ASTM30GraphicViewController {

    private func _setAndAddGraphicBackgroundViewToView(view: UIView) {
        _graphicBackgroundView?.removeFromSuperview()
        _graphicBackgroundForFadeView?.removeFromSuperview()

        func newGraphicView() -> UIImageView {
            let imageView = UIImageView(image: UIImage(named: "GraphicBackground"))
            imageView.frame = view.frame
            imageView.contentMode = .ScaleToFill

            return imageView
        }

        _graphicBackgroundView = newGraphicView()
        _graphicBackgroundForFadeView = newGraphicView()

        view.addSubview(_graphicBackgroundForFadeView!)
        view.addSubview(_graphicBackgroundView!)
    }

    private func _setAndAddGridLayerToView(view: UIView) {
        _graphicBackgroundGridLayer?.removeFromSuperlayer()

        let numberOfGrid: CGFloat = 5

        let interval = CGPoint(
            x: view.frame.size.width/numberOfGrid,
            y: view.frame.size.height/numberOfGrid
        )

        let path = UIBezierPath()

        for i in 1..<Int(numberOfGrid) {
            let linePath = UIBezierPath()

            linePath.moveToPoint(CGPoint(x: 0, y: i.cgFloatValue*interval.y))
            linePath.addLineToPoint(CGPoint(x: view.frame.size.width, y: i.cgFloatValue*interval.y))

            linePath.moveToPoint(CGPoint(x: i.cgFloatValue*interval.x, y: 0))
            linePath.addLineToPoint(CGPoint(x: i.cgFloatValue*interval.x, y: view.frame.size.height))


            path.appendPath(linePath)
        }

        let gridLayer = CAShapeLayer()
        gridLayer.frame = view.frame
        gridLayer.path = path.CGPath

        gridLayer.lineWidth = 2
        gridLayer.strokeColor = UIColor.grayColor().CGColor

        view.layer.addSublayer(gridLayer)

        _graphicBackgroundGridLayer = gridLayer
    }

    private func _setAndAddPointsLayersViewToView(view: UIView) {
        _pointsLinesLayerView?.removeFromSuperview()

        let pointsView = UIView(frame: view.frame)

        view.addSubview(pointsView)

        _pointsLinesLayerView = pointsView
    }

    private func _setAndAddAllPoinsInfoLayersToView(view: UIView) {
        _pointsInfoToLayersDict.values.forEach { layer in layer.removeFromSuperlayer() }

        let allInfos = _pointsInfoToLayersDict.keys
        _pointsInfoToLayersDict.removeAll(keepCapacity: true)

        allInfos.forEach {
            info in
            let shapeLayer = _shapeLayerWithPointsInfo(info)
            _pointsInfoToLayersDict[info] = shapeLayer
            view.layer.addSublayer(shapeLayer)
        }
    }

    private func _setGraphicBackgroundMaskWithPointsInfoName(name: String?) {
        guard let name = name else { return }
        guard let graphicBackgroundView = _graphicBackgroundView else { return }

        let pointsInfoKeyValue = _pointsInfoToLayersDict.filter {
            (info, _) in return info.name == name
        }.first

        MZ.Debugs.assert(pointsInfoKeyValue != nil, "pointsInfo not found with name: \(name)")

        guard let layer = pointsInfoKeyValue?.1 else {
            MZ.Debugs.assertAlwayFalse("layer not found")
            return
        }

        graphicBackgroundView.layer.mask = nil

        _graphicBackgroundMaskLayer = _maskLayerFromLayer(layer)

        testSourceName = name
    }

    private func _setAndAddReferenceToTestSourceArrowsLayerToView(view: UIView) {
        _sourceToReferenceArrowsLayer?.removeFromSuperlayer()

        guard let testSourceName = testSourceName else { return }
        guard let referenceName = referenceName else { return }

        guard let toInfo = poinsInfoWithName(testSourceName) else { return }
        guard let fromInfo = poinsInfoWithName(referenceName) else { return }

        let path = UIBezierPath()

        for fromPoint in fromInfo.points {
            guard let toPoint = toInfo.pointWithKey(fromPoint.key) else {
                MZ.Debugs.assertAlwayFalse("Can not found point with key: \(fromPoint.key)")
                return
            }

            let modifiedFromPoint = _pointFrom(fromPoint.value, inCoordinateSpace: coordinateSpace)
            let modifiedToPoint = _pointFrom(toPoint.value, inCoordinateSpace: coordinateSpace)

            path.appendPath(_arrowPathFromPoint(modifiedFromPoint, toPoint: modifiedToPoint))
        }

        let layer = CAShapeLayer()
        layer.frame = view.frame
        layer.path = path.CGPath
        layer.strokeColor = UIColor.greenColor().CGColor
        layer.lineWidth = 1
        layer.lineCap = "round"
        layer.lineJoin = "bevel"
        view.layer.addSublayer(layer)
        
        _sourceToReferenceArrowsLayer = layer
    }

    private func _shapeLayerWithPointsInfo(pointsInfo: ASTM30PointsInfo) -> CAShapeLayer {
        let points = pointsInfo.points.map { p in return _pointFrom(p.value, inCoordinateSpace: self.coordinateSpace) }

        let path = UIBezierPath()
        path.moveToPoint(points[0])
        points[1..<points.endIndex].forEach { point in path.addLineToPoint(point) }

        if pointsInfo.closePath { path.closePath() }

        let layer = CAShapeLayer()
        layer.path = path.CGPath
        layer.lineWidth = pointsInfo.lineWidth
        layer.strokeColor = pointsInfo.color.CGColor
        layer.fillColor = UIColor.clearColor().CGColor

        let actions: [String: CAAction] = ["strokeColor": NSNull()]
        layer.actions = actions

        return layer
    }

    private func _maskLayerFromLayer(layer: CAShapeLayer) -> CAShapeLayer {
        let maskLayer = CAShapeLayer(layer: layer)
        maskLayer.frame.size = view.frame.size
        maskLayer.path = layer.path

        return maskLayer
    }

    private func _arrowPathFromPoint(from: CGPoint, toPoint to: CGPoint) -> UIBezierPath {
        let lengthOfFromTo = MZ.Maths.distance(p1: from, p2: to)
        let degreesOfFromTo = MZ.Degrees.degressFromP1(from, toP2: to)
        let maxWingsLength = 10.cgFloatValue
        let wingsIntervalDegrees = 30.mzFloatValue

        let path = UIBezierPath()
        let zeroLeft = CGPoint(x: -lengthOfFromTo/2, y: 0)
        let zeroRight = CGPoint(x: lengthOfFromTo/2, y: 0)

        path.moveToPoint(zeroLeft)
        path.addLineToPoint(zeroRight)

        let wingsLength = min(maxWingsLength, lengthOfFromTo/3)
        let wingPoint1 = MZ.Maths.unitVectorFromDegrees(180-wingsIntervalDegrees)*wingsLength
        let wingPoint2 = MZ.Maths.unitVectorFromDegrees(180+wingsIntervalDegrees)*wingsLength

        path.moveToPoint(zeroRight)
        path.addLineToPoint(zeroRight + wingPoint1)
        path.moveToPoint(zeroRight)
        path.addLineToPoint(zeroRight + wingPoint2)

        let centerOfFromTo = CGPoint(x: (to.x + from.x)/2, y: (to.y + from.y)/2)
        let rotationAngle = MZ.Degrees.radiansFromDegrees(degreesOfFromTo).cgFloatValue
        path.applyTransform(CGAffineTransformMakeRotation(rotationAngle))
        path.applyTransform(CGAffineTransformMakeTranslation(centerOfFromTo.x, centerOfFromTo.y))

        return path
    }
}


// MARK: ASTM30GraphicType switch animations
extension ASTM30GraphicViewController {

    private func _animateMaskEnable(maskEnable: Bool, duration: NSTimeInterval) {
        _animateMaskEnableToGraphicBackground(maskEnable, duration: duration)
        _animateMaskEnableToGraphicBackgroundForFade(maskEnable, duration: duration)
        _animateFadeMaskEnableToLayer(_graphicBackgroundGridLayer, maskEnable: maskEnable, duration: duration)
        _animateFadeMaskEnableToLayer(_sourceToReferenceArrowsLayer, maskEnable: maskEnable, duration: duration)
        _animateMaskEnableToPointsLinesLayer(maskEnable, duration: duration)
    }

    private func _animateMaskEnableToGraphicBackground(maskEnable: Bool, duration: NSTimeInterval) {
        guard let view = _graphicBackgroundView else { return }
        guard let maskLayer = _graphicBackgroundMaskLayer else { return }

        if view.layer.mask == nil {
            view.layer.mask = maskLayer
        }

        let maxScale = 5.0
        let minScale = 1.0

        let scale = CAKeyframeAnimation(keyPath: "transform.scale")
        scale.duration = duration
        scale.keyTimes = [0.0, 1.0]
        scale.values = [
            (maskEnable ? maxScale : minScale),
            (maskEnable ? minScale : maxScale),
        ]
        scale.timingFunctions = [CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseInEaseOut)]
        scale.fillMode = kCAFillModeForwards
        scale.removedOnCompletion = false

        maskLayer.removeAllAnimations()
        maskLayer.addAnimation(scale, forKey: "scale")
    }

    private func _animateMaskEnableToGraphicBackgroundForFade(maskEnable: Bool, duration: NSTimeInterval) {
        guard let view = _graphicBackgroundForFadeView else { return }
        view.alpha = maskEnable ? 1 : 0
        UIView.animateWithDuration(duration,
            animations: { view.alpha = maskEnable ? 0 : 1 },
            completion: { _ in self._setGraphicBackgroundWithMaskEnable(maskEnable) }
        )
    }

    private func _animateFadeMaskEnableToLayer(layer: CAShapeLayer?, maskEnable: Bool, duration: NSTimeInterval) {
        guard let layer = layer else { return }

        let fade = CABasicAnimation(keyPath: "opacity")
        fade.fromValue = maskEnable ? 1 : 0
        fade.toValue = maskEnable ? 0 : 1
        fade.duration = duration
        fade.fillMode = kCAFillModeForwards
        fade.removedOnCompletion = false

        layer.removeAllAnimations()
        layer.addAnimation(fade, forKey: "fade")
    }

    private func _animateMaskEnableToPointsLinesLayer(maskEnable: Bool, duration: NSTimeInterval) {
        _pointsInfoToLayersDict.forEach {
            (info, shapeLayer) in
            let nextColorValue = maskEnable == true ? info.colorInMasked : info.color

            guard let nextColor = nextColorValue else { return }

            let color = CABasicAnimation(keyPath: "strokeColor")
            color.fromValue = shapeLayer.strokeColor
            color.toValue = nextColor.CGColor
            color.duration = duration
            
            shapeLayer.addAnimation(color, forKey: "color")
        }
    }
}


// MARK: Supports
extension ASTM30GraphicViewController {

    private func _pointFrom(point: CGPoint, inCoordinateSpace coordinateSpace: ASTM30CoordinateSpace)
    -> CGPoint {
        let realX = view.frame.width*((point.x - coordinateSpace.xMin)/coordinateSpace.xLength)
        let realY = view.frame.height - view.frame.height*((point.y - coordinateSpace.yMin)/coordinateSpace.yLength)

        return CGPoint(x: realX, y: realY)
    }

    // useless now
    private func _colorAtPoint(point:CGPoint) -> UIColor {
        let colorSpace:CGColorSpace = CGColorSpaceCreateDeviceRGB()!
        let bitmapInfo = CGBitmapInfo(rawValue: CGImageAlphaInfo.PremultipliedLast.rawValue)
        var pixelData:[UInt8] = [0, 0, 0, 0]

        let context = CGBitmapContextCreate(&pixelData, 1, 1, 8, 4, colorSpace, bitmapInfo.rawValue)!
        CGContextTranslateCTM(context, -point.x, -point.y);
        view.layer.renderInContext(context)

        let red:CGFloat = CGFloat(pixelData[0])/CGFloat(255.0)
        let green:CGFloat = CGFloat(pixelData[1])/CGFloat(255.0)
        let blue:CGFloat = CGFloat(pixelData[2])/CGFloat(255.0)
        let alpha:CGFloat = CGFloat(pixelData[3])/CGFloat(255.0)

        let color:UIColor = UIColor(red: red, green: green, blue: blue, alpha: alpha)
        return color
    }
}
